tsl/src/remote/dist_ddl.c

```diff
@@ -421,20 +447,83 @@ dist_ddl_process_rename(ProcessUtilityArgs *args)
 		 * Save oid for dist_ddl_end execution.
 		 */
 		dist_ddl_state.relid = linitial_oid(args->hypertable_list);
-		dist_ddl_state_set_exec_type(DIST_DDL_EXEC_ON_END);
+		dist_ddl_state_schedule(DIST_DDL_EXEC_ON_END, args);
 		return;
 	}
 
 	/* Only handles other renamings here (e.g., triggers) */
 	if (dist_ddl_state_set_hypertable(args))
-		dist_ddl_state_set_exec_type(DIST_DDL_EXEC_ON_START);
+		dist_ddl_state_schedule(DIST_DDL_EXEC_ON_START, args);
 }
 
 static void
 dist_ddl_process_grant_on_database(ProcessUtilityArgs *args)
 {
-	(void) args;
-	/* TBD */
+	GrantStmt *stmt = castNode(GrantStmt, args->parsetree);
+	const char *dbname;
+	bool dbmatch;
+	List *cmd_descriptors = NIL;
+	ListCell *i, *j;
+
+	if (dist_util_membership() != DIST_MEMBER_ACCESS_NODE)
+		return;
+
+	/* Check if the current database is used */
+	dbname = get_database_name(MyDatabaseId);
+	dbmatch = false;
+	foreach (i, stmt->objects)
+	{
+		Value *value = lfirst(i);
+
+		if (!strcmp(dbname, strVal(value)))
+		{
+			dbmatch = true;
+			break;
+		}
+	}
+
+	if (!dbmatch)
+		return;
+
+	/* Prevent using and mixing several databases in the statement */
+	if (list_length(stmt->objects) > 1)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("operation not supported on distributed database")));
```

Comment:
We can't support multiple distributed databases, because all of them have a different set of data nodes at least. Mixing with local databases might work, but we don't know if they are distributed or not. Same applies for the `WARNING` message, it is unclear when it should be issued. It will end up showing warning message for every `GRANT ON DATABASE` on the access node, do we want that?

Not sure what you mean by not blocking it. This operation will not work, unless you want blindly include other databases in the GRANT command to the data nodes. Is that the idea?

It should not affect single node, because we check the node status first specifically to be the access node.

Reply:
Now we block GRANTing on multiple databases if the current database is distributed.

But, we don't block GRANTing on multiple database if the current database is _not_ distributed, but a database in the statement _is_ distributed. So, in that case, we can issue GRANTS on a distributed database from a non-distributed database and we will not forward the GRANT to the data nodes.

So, I see an inconsistency in how we block things. In fact, I am not sure it is possible to fix this properly, so that's why I am wondering if we shouldn't even try to fix it.